src/python/pants/base/specs_parser.py

```diff
@@ -109,9 +119,31 @@ def parse_specs(
         convert_dir_literal_to_address_literal: bool,
         unmatched_glob_behavior: GlobMatchErrorBehavior = GlobMatchErrorBehavior.error,
     ) -> Specs:
-        return Specs.create(
-            (self.parse_spec(spec) for spec in specs),
+        include_specs = []
+        ignore_specs = []
+        for spec_str in specs:
+            spec, is_ignore = self.parse_spec(spec_str)
+            if is_ignore:
+                ignore_specs.append(spec)
+            else:
+                include_specs.append(spec)
+
+        includes = RawSpecs.create(
+            include_specs,
             convert_dir_literal_to_address_literal=convert_dir_literal_to_address_literal,
             unmatched_glob_behavior=unmatched_glob_behavior,
             filter_by_global_options=True,
         )
+        ignores = RawSpecs.create(
+            ignore_specs,
+            convert_dir_literal_to_address_literal=convert_dir_literal_to_address_literal,
+            unmatched_glob_behavior=unmatched_glob_behavior,
+            # By setting the below to False, we will end up matching some targets
+            # that cannot have been resolved by the include specs. For example, if the user runs
+            # `--filter-target-type=my_tgt :: !dir::`, the ignores may match targets that are not
+            # my_tgt. However, there also is no harm in over-matching with ignores. Setting
+            # this to False (over-conservatively?) ensures that if a user says to ignore
+            # something, we definitely do.
+            filter_by_global_options=False,
+        )
+        return Specs(includes, ignores)
```

Comment:
I would expect to see some functionality to subtract the ignores from the includes to get the final set of specs to operate on, or is that somewhere else I overlooked?